# 微分可能ビリーフロパゲーション復号法

In [1]:
import alist_loader
import jax.numpy as jnp

## 本家のリポジトリと同じ.alistファイルを使う
- リーダーも適当に自作

In [2]:
filename = "../../DU-Book/Chapter_5/6.3.alist"

In [3]:
H, U, V = alist_loader.load_alist(filename)

n = 6, m = 3
cmap = 2, rmap = 3


## 微分可能ビリーフプロパゲーション復号法

In [4]:
esize = U.shape[-1]

In [5]:
var = 0.794328
Lambda = 2.0*jnp.array([1.620803, 0.264281, -0.031637, 
    -0.127654, 0.746347, 1.003543])/var
alpha = jnp.zeros(esize) 
beta = jnp.zeros(esize)

In [6]:
def bmod(x):
    return x - 2*jnp.floor(x/2)

In [7]:
for i in range(1,4):
    beta = (U.T@U-jnp.eye(len(U.T))) @ alpha + U.T@Lambda
    tmp = jnp.exp((V.T@V-jnp.eye(len(U.T))) @ jnp.log(jnp.abs(jnp.tanh(beta/2))))
    alpha_abs = 2*jnp.arctanh(tmp)
    tmp = 1 -2*V.T @ bmod(V@((-jnp.sign(beta) + 1)/2))
    alpha_sign = tmp * jnp.sign(beta)
    alpha = alpha_sign * alpha_abs
gamma = U@alpha + Lambda

In [8]:
print(gamma)

[4.397419  1.6925247 1.7110999 1.7111    2.083964  2.703277 ]


## 対数事後確率比の答え合わせ

上記の設定において、対数事後確率比をBP復号法により計算すると

4.397419
1.692525
1.711101
1.711101
2.083965
2.703277

となるらしい(和田山著「誤り訂正技術の基礎」演習問題14.4)。
上記の計算結果は十分な精度で一致している。